torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -0,0 +1,442 @@
+    {at::kInt, ncclInt32},
+    {at::kLong, ncclInt64},
+    {at::kHalf, ncclHalf},
+};
+
+// Helper function that gets the data type and issues error if not supported
+ncclDataType_t getNcclDataType(at::ScalarType type) {
+  try {
+    return ncclDataType.at(type);
+  } catch (std::out_of_range& e) {
+    throw std::runtime_error("Unsupported data type for NCCL process group");
+  }
+}
+
+// Helper function that gets the device list to determine the CUDA devices
+std::vector<int> getDevicesFromKey(const std::string& deviceSeq) {
+  std::stringstream ss(deviceSeq);
+  std::string device;
+  std::vector<int> devices;
+  while (std::getline(ss, device, ',')) {
+    devices.push_back(stoi(device));
+  }
+  return devices;
+}
+
+// Get the deviceList String from the list of devices
+std::string getKeyFromDevices(const std::vector<int>& devices) {
+  std::string deviceList;
+  for (auto device : devices) {
+    if (deviceList.empty()) {
+      deviceList = std::to_string(device);
+    } else {
+      deviceList += "," + std::to_string(device);
+    }
+  }
+  return deviceList;
+}
+
+// Get the list of devices from list of tensors
+std::vector<int> getDevicesOfTensors(const std::vector<at::Tensor>& tensors) {
+  std::vector<int> res;
+  for (auto& tensor : tensors) {
+    res.push_back(tensor.get_device());
+  }
+  return res;
+}
+
+// Helper that lets the input ncclStreams to wait for the THC stream
+void syncStreams(
+    THCState* thcState,
+    const std::vector<int>& devices,
+    std::vector<CUDAEvent>& ncclEvents,
+    std::vector<CUDAStream>& ncclStreams) {
+  CUDADevice gpuGuard;
+  for (auto i = 0; i < devices.size(); ++i) {
+    gpuGuard.setDevice(devices[i]);
+    auto currentThcStream = THCState_getCurrentStream(thcState);
+    CUDAStream& ncclStream = ncclStreams[i];
+    CUDAEvent& ncclEvent = ncclEvents[i];
+
+    C10D_CUDA_CHECK(cudaEventRecord(ncclEvent.getEvent(), currentThcStream));
+    C10D_CUDA_CHECK(
+        cudaStreamWaitEvent(ncclStream.getStream(), ncclEvent.getEvent(), 0));
+  }
+}
+
+} // namespace
+
+ProcessGroupNCCL::WorkNCCL::WorkNCCL(const std::vector<int>& devices)
+    : devices_(devices) {
+  // GPU deviceGuard
+  CUDADevice gpuGuard;
+  cudaEvents_.resize(devices.size());
+  // Now create the CUDA events
+  for (auto i = 0; i < devices.size(); ++i) {
+    // Guard the GPU
+    gpuGuard.setDevice(devices[i]);
+    cudaEvents_[i] = CUDAEvent::create(cudaEventDisableTiming);
+  }
+}
+
+ProcessGroupNCCL::WorkNCCL::~WorkNCCL() {}
+
+// Checking the work's corresponding CUDA events' status
+bool ProcessGroupNCCL::WorkNCCL::isCompleted() const {
+  CUDADevice gpuGuard;
+  for (auto i = 0; i < devices_.size(); ++i) {
+    // Get to the device
+    gpuGuard.setDevice(devices_[i]);
+    auto& cudaEvent = cudaEvents_[i];
+    // Query the event
+    auto ret = cudaEventQuery(cudaEvent.getEvent());
+    if (ret != cudaSuccess && ret != cudaErrorNotReady) {
+      // Throw directly
+      C10D_CUDA_CHECK(ret);
```

Comment:
The 3 comments here don't add value IMO; they just describe the function call below